mlir/lib/Dialect/Rock/Transforms/RemoveRedundantCasts.cpp

```diff
@@ -0,0 +1,510 @@
+    auto outputType = cast<RankedTensorType>(output->getResult(0).getType());
+
+    // Verify that the output element type is smaller than the input
+    if (isa<IntegerType>(inputType.getElementType()) &&
+        isa<IntegerType>(outputType.getElementType())) {
+      assert(cast<IntegerType>(outputType.getElementType()).getWidth() < 32 &&
+             "Output element type must be smaller than i32");
+    } else if (isa<FloatType>(inputType.getElementType()) &&
+               isa<FloatType>(outputType.getElementType())) {
+      assert(cast<FloatType>(outputType.getElementType()).getWidth() < 32 &&
+             "Output element type must be smaller than f32");
+    } else {
+      llvm_unreachable("Unsupported element type for truncation");
+    }
+
+    // Create indexing maps - both input and output use the same identity map
+    MLIRContext *ctx = rewriter.getContext();
+    int64_t rank = inputType.getRank();
+    SmallVector<AffineExpr> exprs;
+    for (int64_t i = 0; i < rank; ++i) {
+      exprs.push_back(getAffineDimExpr(i, ctx));
+    }
+    auto identityMap = AffineMap::get(rank, 0, exprs, ctx);
+    SmallVector<AffineMap> indexingMaps = {identityMap, identityMap};
+
+    // Create iterator types - all parallel
+    SmallVector<utils::IteratorType> iteratorTypes(
+        rank, utils::IteratorType::parallel);
+
+    // Create the linalg.generic operation
+    auto loc = input->getLoc();
+    auto genericOp = rewriter.create<linalg::GenericOp>(
+        loc,
+        /*resultTypes=*/TypeRange{outputType},
+        /*inputs=*/ValueRange{input->getResult(0)},
+        /*outputs=*/ValueRange{output->getResult(0)},
+        /*indexingMaps=*/indexingMaps,
+        /*iteratorTypes=*/iteratorTypes,
+        /*doc=*/"",
+        /*library_call=*/"",
+        /*bodyBuild=*/
+        [](OpBuilder &builder, Location loc, ValueRange args) {
+          Value blockArg = args[0];
+          Value outputArg = args[1];
+          Type oType = outputArg.getType();
+          Value truncResult =
+              builder.create<arith::TruncFOp>(loc, oType, blockArg);
+          builder.create<linalg::YieldOp>(loc, truncResult);
+        });
+
+    return genericOp;
+  }
+
+  bool handleRockGemmWrapper(Operation *rockOp, linalg::GenericOp extGen,
+                             PatternRewriter &rewriter) const {
+    LLVM_DEBUG(llvm::dbgs()
+               << "\tFound "
+               << "RockGemmWrapperInterface/RockGemmGemmWrapperInterface: "
+               << *rockOp << "\n");
+
+    Value extGenResult = extGen.getResult(0);
+    Value extGenInput = extGen.getInputs()[0];
+
+    auto extOutputType = dyn_cast<RankedTensorType>(extGenResult.getType());
+    auto extInputType = dyn_cast<RankedTensorType>(extGenInput.getType());
+
+    if (!extOutputType || !extInputType) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\tOne or more types are not RankedTensorType, skipping\n");
+      return false;
+    }
```

Comment:
You should just be checking for ShapedType instead of RankedTensorType. It is better and handles more general type of containers/buffers.